File: src/utils/chunkingService.js
/**
 * Chunking Service
 * Splits documents into overlapping chunks with sentence-boundary awareness.
 */

const DEFAULT_CHUNK_SIZE = parseInt(process.env.CHUNK_SIZE, 10) || 1000;   // words
const DEFAULT_CHUNK_OVERLAP = parseInt(process.env.CHUNK_OVERLAP, 10) || 50; // words

/**
 * Split text into overlapping, sentence-aware chunks.
 * @param {string} text - The full document text
 * @param {Object} opts
 * @param {number} opts.chunkSize   - Max words per chunk (default 1000)
 * @param {number} opts.chunkOverlap - Overlap in words (default 50)
 * @returns {string[]} Array of text chunks
 */
function chunkText(text, opts = {}) {
  const chunkSize = opts.chunkSize || DEFAULT_CHUNK_SIZE;
  const chunkOverlap = opts.chunkOverlap || DEFAULT_CHUNK_OVERLAP;

  if (!text || text.trim().length === 0) {
    return [];
  }

  // Split text into sentences first for boundary awareness
  const sentences = splitIntoSentences(text);

  const chunks = [];
  let currentChunk = [];
  let currentWordCount = 0;

  for (const sentence of sentences) {
    const sentenceWords = countWords(sentence);

    // If a single sentence exceeds chunk size, force-split it by words
    if (sentenceWords > chunkSize) {
      // Flush what we have
      if (currentChunk.length > 0) {
        chunks.push(currentChunk.join(' '));
        currentChunk = getOverlapSentences(currentChunk, chunkOverlap);
        currentWordCount = countWords(currentChunk.join(' '));
      }
      // Split the long sentence by words
      const forcedChunks = forceSplitByWords(sentence, chunkSize, chunkOverlap);
      chunks.push(...forcedChunks);
      currentChunk = [];
      currentWordCount = 0;
      continue;
    }

    // If adding this sentence would exceed chunk size, start a new chunk
    if (currentWordCount + sentenceWords > chunkSize && currentChunk.length > 0) {
      chunks.push(currentChunk.join(' '));
      // Keep overlap from the end of the current chunk
      currentChunk = getOverlapSentences(currentChunk, chunkOverlap);
      currentWordCount = countWords(currentChunk.join(' '));
    }

    currentChunk.push(sentence);
    currentWordCount += sentenceWords;
  }

  // Flush remaining
  if (currentChunk.length > 0) {
    const remaining = currentChunk.join(' ').trim();
    if (remaining.length > 0) {
      chunks.push(remaining);
    }
  }

  return chunks;
}

/**
 * Split text into sentences using common delimiters.
 */
function splitIntoSentences(text) {
  // Split on sentence-ending punctuation followed by whitespace
  const raw = text.split(/(?<=[.!?])\s+/);
  return raw.map(s => s.trim()).filter(s => s.length > 0);
}

/**
 * Count words in a string.
 */
function countWords(text) {
  if (!text || text.trim().length === 0) return 0;
  return text.trim().split(/\s+/).length;
}

/**
 * Get sentences from the end of a chunk that fit within the overlap word count.
 */
function getOverlapSentences(sentences, overlapWords) {
  const result = [];
  let wordCount = 0;

  for (let i = sentences.length - 1; i >= 0; i--) {
    const sentWords = countWords(sentences[i]);
    if (wordCount + sentWords > overlapWords) break;
    result.unshift(sentences[i]);
    wordCount += sentWords;
  }

  return result;
}

/**
 * Force-split a very long sentence by word count.
 */
function forceSplitByWords(text, chunkSize, overlap) {
  const words = text.split(/\s+/);
  const chunks = [];
  let start = 0;

  while (start < words.length) {
    const end = Math.min(start + chunkSize, words.length);
    chunks.push(words.slice(start, end).join(' '));
    start = end - overlap;
    if (start >= words.length) break;
    if (end === words.length) break;
  }

  return chunks;
}

/**
 * Create chunk metadata objects for storage.
 * @param {string} documentId - Unique document identifier
 * @param {string} filename - Original filename
 * @param {string[]} chunks - Array of text chunks
 * @returns {Object[]} Array of { chunkId, documentId, filename, chunkIndex, totalChunks, text, wordCount }
 */
function createChunkMetadata(documentId, filename, chunks) {
  return chunks.map((text, index) => ({
    chunkId: `${documentId}_chunk_${index}`,
    documentId,
    filename,
    chunkIndex: index,
    totalChunks: chunks.length,
    text,
    wordCount: countWords(text)
  }));
}

module.exports = {
  chunkText,
  createChunkMetadata,
  countWords,
  splitIntoSentences,
  DEFAULT_CHUNK_SIZE,
  DEFAULT_CHUNK_OVERLAP
};

File: app/utils/chunking_service.py
"""
Chunking Service
Splits documents into overlapping chunks with sentence-boundary awareness.
"""
import os
import re
from typing import Dict, List, Optional

DEFAULT_CHUNK_SIZE = int(os.getenv("CHUNK_SIZE", "500"))        # characters
DEFAULT_CHUNK_OVERLAP = int(os.getenv("CHUNK_OVERLAP", "100"))  # characters


def chunk_text(
    text: str,
    chunk_size: Optional[int] = None,
    chunk_overlap: Optional[int] = None,
) -> List[str]:
    """
    Split *text* into overlapping, sentence-aware chunks.

    Args:
        text: Full document text.
        chunk_size: Max characters per chunk (default from env / 500).
        chunk_overlap: Overlap in characters (default from env / 100).

    Returns:
        List of text chunks.
    """
    size = chunk_size or DEFAULT_CHUNK_SIZE
    overlap = chunk_overlap or DEFAULT_CHUNK_OVERLAP

    if not text or not text.strip():
        return []

    # Validate: overlap must be smaller than chunk size
    if overlap >= size:
        print(f"[Chunking] Overlap ({overlap}) >= chunkSize ({size}). "
              f"Clamping to {int(size * 0.2)}.")
        overlap = int(size * 0.2)

    sentences = _split_into_sentences(text)

    chunks: List[str] = []
    current_chunk: List[str] = []
    current_len = 0

    for sentence in sentences:
        s_len = len(sentence)

        # Single sentence exceeds chunk size → force-split by characters
        if s_len > size:
            if current_chunk:
                chunks.append(" ".join(current_chunk))
                current_chunk = _get_overlap_sentences(current_chunk, overlap)
                current_len = sum(len(s) for s in current_chunk) + max(len(current_chunk) - 1, 0)
            chunks.extend(_force_split(sentence, size, overlap))
            current_chunk = []
            current_len = 0
            continue

        # Adding this sentence would exceed chunk size → start new chunk
        sep_len = 1 if current_chunk else 0  # space separator
        if current_len + sep_len + s_len > size and current_chunk:
            chunks.append(" ".join(current_chunk))
            current_chunk = _get_overlap_sentences(current_chunk, overlap)
            current_len = sum(len(s) for s in current_chunk) + max(len(current_chunk) - 1, 0)

        current_chunk.append(sentence)
        current_len += (1 if current_len > 0 else 0) + s_len

    # Flush remaining
    if current_chunk:
        remaining = " ".join(current_chunk).strip()
        if remaining:
            if chunks and len(remaining) < overlap:
                chunks[-1] += " " + remaining
            else:
                chunks.append(remaining)

    return chunks


def create_chunk_metadata(
    document_id: str,
    filename: str,
    chunks: List[str],
) -> List[Dict]:
    """Build metadata dicts for each chunk."""
    total = len(chunks)
    return [
        {
            "chunk_id": f"{document_id}_chunk_{i}",
            "document_id": document_id,
            "filename": filename,
            "chunk_index": i,
            "total_chunks": total,
            "text": chunk,
            "word_count": _count_words(chunk),
        }
        for i, chunk in enumerate(chunks)
    ]


# ─── Helpers ──────────────────────────────────────────────

def _split_into_sentences(text: str) -> List[str]:
    parts = re.split(r"(?<=[.!?])\s+", text)
    return [s.strip() for s in parts if s.strip()]


def _count_words(text: str) -> int:
    if not text or not text.strip():
        return 0
    return len(text.strip().split())


def _get_overlap_sentences(sentences: List[str], overlap_chars: int) -> List[str]:
    result: List[str] = []
    char_count = 0
    for s in reversed(sentences):
        sl = len(s)
        if char_count + sl > overlap_chars:
            break
        result.insert(0, s)
        char_count += sl
    return result


def _force_split(text: str, size: int, overlap: int) -> List[str]:
    # Safety: prevent infinite loop if bad params slip through
    if size <= 0:
        size = 500
    if overlap >= size:
        overlap = int(size * 0.2)

    chunks: List[str] = []
    start = 0
    while start < len(text):
        end = min(start + size, len(text))
        chunk = text[start:end].strip()
        if chunk:
            chunks.append(chunk)
        # Advance by (size - overlap), minimum 1 char to guarantee progress
        step = max(size - overlap, 1)
        start += step
    return chunks
